### DISCUSSION TITLE
Configurate using code

### DISCUSSION THREAD

**Discussion by NightWinkle** [facebookresearch/hydra#2805] [Q&A]
Is there a way to configurate using python code input ? 
I'm trying to configurate experiments by defining an input function f(x).
I've been thinking of the following solution : 

1. Create a input.py file containing a class : 
```py
class Input:
    def __init__(self, param1):
        self.param1 = param1
    def f(self, x):
        return x*self.param1 / 2
```
2. Instantiate using yaml config : 
```yaml
input:
    _target_: input.Input
    param1: 0.3
```
3. Add config path to pythonpath
```py
hcfg = HydraConfig.get()
for source in hcfg.runtime.config_sources:
    os.environ["PYTHONPATH"] = os.environ.get("PYTHONPATH", "") + ":" + source.path
```

However, this looks like a quite dirty solution. Is there any better way ?

**Comment by odelalleau** [comment]
What do you mean "configurate"? Is the problem that your `input.py` is not in `PYTHONPATH`? If yes AFAIK it's your responsibility to ensure that whatever you instantiate using the `_target_` mechanism is in `PYTHONPATH`. In general Hydra has no way to know where those are located.

**Reply by NightWinkle** [reply]
I want to specify functions as input parameters. As an example usecase, if you are using Hydra to configurate PDEs solving, you might want to be able to specify boundary conditions (initial value function).

**Reply by odelalleau** [reply]
Ok -- it sounds like Hydra's instantiation mechanism is indeed what you should be using then. What is it exactly you would like to do differently? If what you want is to get rid of the `.yaml`, you can use [structured configs](https://hydra.cc/docs/tutorials/structured_config/intro/) to do everything in Python. You can also look into [hydra-zen](https://github.com/mit-ll-responsible-ai/hydra-zen), which offers additional functionalities to automate the generation of structured configs from Python code.

**Reply by NightWinkle** [reply]
Sorry for the late response.
What I mean is I want to have "code" as parameter. For instance I would like one of the input of my program to be a function that the user can define (and I'm perfectly fine if it is using yaml).
For instance, my yaml would contain :
```yaml
source:
    f: x**2 + 2*t
``` 
Which then allows me to configurate simulations whose input is a matrix `f(x[:, None], t[None, :])`. From my understanding of previous issues and discussions here, this is however not possible. This is why I was suggesting as an alternative that I might define a function f in a python file next to my yaml config, and then instantiate this python file using the yaml config. 
However, this is actually not working because the python file is not in python path.
As I understand there is no dedicated solution for this kind of problem ?

**Reply by odelalleau** [reply]
Thanks for the clarification. One way to achieve this is through resolvers. This would become for instance
```yaml
source:
  f: "${func: x**2 + 2*t}"
```
with
```python
OmegaConf.register_new_resolver("func", lambda txt: eval(f"lambda x, t: {txt}"))
```

Just be aware of the security risks associated with the usage of `eval()` in Python.
